godot-core/src/builtin/signal/futures.rs

```diff
@@ -0,0 +1,313 @@
+/*
+ * Copyright (c) godot-rust; Bromeon and contributors.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/.
+ */
+
+use core::panic;
+use std::fmt::Display;
+use std::future::{Future, IntoFuture};
+use std::pin::Pin;
+use std::sync::{Arc, Mutex};
+use std::task::{Context, Poll, Waker};
+
+use crate::builtin::{Callable, RustCallable, Variant};
+use crate::classes::object::ConnectFlags;
+use crate::meta::ParamTuple;
+use crate::obj::{EngineBitfield, WithBaseField};
+use crate::registry::signal::TypedSignal;
+
+use super::Signal;
+
+/// The panicking counter part to the [`TrySignalFuture`].
+///
+/// This future works in the same ways as the TrySignalFuture, but panics when the signal object is freed instead of resolving to an
+/// [`Result::Err`]
+pub struct SignalFuture<R: ParamTuple + Sync + Send>(TrySignalFuture<R>);
+
+impl<R: ParamTuple + Sync + Send> SignalFuture<R> {
+    fn new(signal: Signal) -> Self {
+        Self(TrySignalFuture::new(signal))
+    }
+}
+
+impl<R: ParamTuple + Sync + Send> Future for SignalFuture<R> {
+    type Output = R;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let poll_result = self.get_mut().0.poll(cx);
+
+        match poll_result {
+            Poll::Pending => Poll::Pending,
+            Poll::Ready(Ok(value)) => Poll::Ready(value),
+            Poll::Ready(Err(TrySignalFutureError)) => panic!(
+                "The signal object of a SignalFuture was freed while the future was still waiting for the signal to be emitted!"
+            ),
+        }
+    }
+}
+
+// Not derived, otherwise an extra bound `Output: Default` is required.
+struct SignalFutureData<T> {
+    state: TrySignalFutureState<T>,
+    waker: Option<Waker>,
+}
+
+impl<T> Default for SignalFutureData<T> {
+    fn default() -> Self {
+        Self {
+            state: Default::default(),
+            waker: None,
+        }
+    }
+}
```

Comment:
The purpose of `SignalFutureData` isn't quite clear to me. Could you add some RustDoc on it?

Also, might there be a better name for this than "data"?


In the `Default` impl, the field `state` having an explicit value would make it a bit easier to understand.

Reply:
I very much dislike the name. The data is the shared data between the future and the resolver. This encompasses both the state enum `SignalFutureState` and the optional `Waker`.